blockchain/skipchain/validator.go

```diff
@@ -0,0 +1,79 @@
+package skipchain
+
+import (
+	"bytes"
+
+	proto "github.com/golang/protobuf/proto"
+	"go.dedis.ch/fabric"
+	"go.dedis.ch/fabric/blockchain"
+	"go.dedis.ch/fabric/consensus"
+	"go.dedis.ch/fabric/encoding"
+	"golang.org/x/xerrors"
+)
+
+// blockValidator is a validator for incoming protobuf messages to decode them
+// into a skipblock and for commit queued blocks when appropriate.
+//
+// - implements consensus.Validator
+type blockValidator struct {
+	*Skipchain
+
+	validator blockchain.Validator
+	buffer    SkipBlock
+}
+
+// Validate implements consensus.Validator. It decodes the message into a block
+// and validate its integrity. It returns the block if it is correct, otherwise
+// the error.
+func (v *blockValidator) Validate(pb proto.Message) (consensus.Proposal, error) {
+	factory := v.GetBlockFactory().(blockFactory)
+
+	block, err := factory.decodeBlock(pb)
+	if err != nil {
+		return nil, encoding.NewDecodingError("block", err)
+	}
+
+	genesis, err := v.db.Read(0)
+	if err != nil {
+		return nil, xerrors.Errorf("couldn't read genesis block: %v", err)
+	}
+
+	if !bytes.Equal(genesis.GetHash(), block.GenesisID.Bytes()) {
+		return nil, xerrors.Errorf("mismatch genesis hash %v != %v", genesis.hash, block.GenesisID)
```

Comment:
If it happens that the hash is empty, the error message can be misleading:
"mismatch genesis hash  != aef123.."
```suggestion
		return nil, xerrors.Errorf("mismatch genesis hash '%v' != '%v'", genesis.hash, block.GenesisID)
```